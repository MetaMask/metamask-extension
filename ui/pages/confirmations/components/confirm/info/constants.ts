export const UNLIMITED_THRESHOLD = 10 ** 15;
