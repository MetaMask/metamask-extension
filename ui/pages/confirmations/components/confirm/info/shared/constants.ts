export const HEX_ZERO = '0x0';

export const TOKEN_VALUE_UNLIMITED_THRESHOLD = 10 ** 15;
