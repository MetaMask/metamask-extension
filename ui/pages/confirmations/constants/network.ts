import { BtcAccountType, KeyringAccountType } from '@metamask/keyring-api';

// This map is used to display the account type label next to the token / account name
// Add more account type here to support labels in case we need them for other networks
export const accountTypeLabel: Partial<Record<KeyringAccountType, string>> = {
  [BtcAccountType.P2pkh]: 'Legacy',
  [BtcAccountType.P2sh]: 'Nested SegWit',
  [BtcAccountType.P2wpkh]: 'Native SegWit',
  [BtcAccountType.P2tr]: 'Taproot',
};
