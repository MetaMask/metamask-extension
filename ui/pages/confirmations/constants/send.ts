export enum SendPages {
  ASSET = 'asset',
  AMOUNT = 'amount',
  LOADER = 'loader',
  RECIPIENT = 'recipient',
}
