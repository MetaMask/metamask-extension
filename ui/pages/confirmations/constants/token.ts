export const ERC20_DEFAULT_DECIMALS = 18;
