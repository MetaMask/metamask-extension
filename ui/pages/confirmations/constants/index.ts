import { CHAINLIST_CHAIN_IDS_MAP } from '../../../../shared/constants/network';

export const IGNORE_GAS_LIMIT_CHAIN_IDS = [CHAINLIST_CHAIN_IDS_MAP.MANTLE];

export const TYPED_SIGNATURE_VERSIONS = {
  V1: 'V1',
  V3: 'V3',
  V4: 'V4',
};

export const SPENDING_CAP_UNLIMITED_MSG = 'UNLIMITED MESSAGE';

export const TypedSignSignaturePrimaryTypes = {
  PERMIT: 'Permit',
  ORDER: 'Order',
};
