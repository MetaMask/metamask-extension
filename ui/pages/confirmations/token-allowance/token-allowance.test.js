import React from 'react';
import configureMockStore from 'redux-mock-store';
import { act, fireEvent } from '@testing-library/react';
import thunk from 'redux-thunk';
import { EthAccountType } from '@metamask/keyring-api';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import { KeyringType } from '../../../../shared/constants/keyring';
import { ETH_EOA_METHODS } from '../../../../shared/constants/eth-methods';
import { mockNetworkState } from '../../../../test/stub/networks';
import { CHAIN_IDS } from '../../../../shared/constants/network';
import TokenAllowance from './token-allowance';

const testTokenAddress = '0xC011a73ee8576Fb46F5E1c5751cA3B9Fe0af2a6F';
const state = {
  appState: {
    customTokenAmount: '1',
  },
  metamask: {
    ...mockNetworkState({ chainId: CHAIN_IDS.MAINNET }),
    accounts: {
      '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc': {
        address: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
        balance: '0x0',
      },
    },
    gasEstimateType: 'none',
    internalAccounts: {
      accounts: {
        'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3': {
          address: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
          id: 'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3',
          metadata: {
            name: 'Account 1',
            keyring: {
              type: 'HD Key Tree',
            },
          },
          options: {},
          methods: ETH_EOA_METHODS,
          type: EthAccountType.Eoa,
        },
        '07c2cfec-36c9-46c4-8115-3836d3ac9047': {
          address: '0xc42edfcc21ed14dda456aa0756c153f7985d8813',
          id: '07c2cfec-36c9-46c4-8115-3836d3ac9047',
          metadata: {
            name: 'Account 2',
            keyring: {
              type: 'HD Key Tree',
            },
          },
          options: {},
          methods: ETH_EOA_METHODS,
          type: EthAccountType.Eoa,
        },
      },
      selectedAccount: 'cf8dace4-9439-4bd4-b3a8-88c821c8fcb3',
    },
    accountsByChainId: {
      '0x1': {
        '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc': {
          address: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
          balance: '0x0',
        },
      },
    },
    addressBook: [
      {
        address: '0xc42edfcc21ed14dda456aa0756c153f7985d8813',
        chainId: '0x5',
        isEns: false,
        memo: '',
        name: 'Address Book Account 1',
      },
    ],
    ...mockNetworkState({ chainId: CHAIN_IDS.MAINNET, nickname: 'mainnet' }),
    preferences: {
      showFiatInTestnets: true,
    },
    currencyRates: {},
    knownMethodData: {},
    tokens: [
      {
        address: testTokenAddress,
        symbol: 'SNX',
        decimals: 18,
        image: 'testImage',
        isERC721: false,
      },
      {
        address: '0xaD6D458402F60fD3Bd25163575031ACDce07538U',
        symbol: 'DAU',
        decimals: 18,
        image: null,
        isERC721: false,
      },
    ],
    transactions: [],
    keyrings: [
      {
        type: KeyringType.hdKeyTree,
        accounts: ['0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc'],
      },
    ],
    nextNonce: 1,
    customNonceValue: '',
    pendingApprovals: {
      '741bad30-45b6-11ef-b6ec-870d18dd6c01': {
        id: '741bad30-45b6-11ef-b6ec-870d18dd6c01',
        origin: 'http://127.0.0.1:8080',
        type: 'transaction',
        time: 1721383540624,
        requestData: {
          txId: '741bad30-45b6-11ef-b6ec-870d18dd6c01',
        },
        requestState: null,
        expectsResult: true,
      },
    },
  },
  history: {
    mostRecentOverviewPage: '/',
  },
  confirmTransaction: {
    txData: {},
  },
  send: {
    draftTransactions: {},
  },
};

const mockShowModal = jest.fn();
const mockedState = jest.mocked(state);

jest.mock('../../../store/actions', () => ({
  getGasFeeTimeEstimate: jest.fn().mockImplementation(() => Promise.resolve()),
  gasFeeStartPollingByNetworkClientId: jest
    .fn()
    .mockResolvedValue('pollingToken'),
  gasFeeStopPollingByPollingToken: jest.fn(),
  addPollingTokenToAppState: jest.fn(),
  removePollingTokenFromAppState: jest.fn(),
  getNetworkConfigurationByNetworkClientId: jest.fn().mockImplementation(() =>
    Promise.resolve({
      chainId: '0x5',
    }),
  ),
  updateTransactionGasFees: () => ({ type: 'UPDATE_TRANSACTION_PARAMS' }),
  updatePreviousGasParams: () => ({ type: 'UPDATE_TRANSACTION_PARAMS' }),
  createTransactionEventFragment: jest.fn(),
  getNextNonce: () => jest.fn(),
  showModal: () => mockShowModal,
  updateCustomNonce: () => ({ type: 'UPDATE_TRANSACTION_PARAMS' }),
  estimateGas: jest.fn().mockImplementation(() => Promise.resolve()),
}));

jest.mock('../../../contexts/gasFee', () => ({
  useGasFeeContext: () => ({
    maxPriorityFeePerGas: '0.1',
    maxFeePerGas: '0.1',
    updateTransaction: jest.fn(),
  }),
}));

jest.mock('react-router-dom', () => {
  const original = jest.requireActual('react-router-dom');
  return {
    ...original,
    useHistory: () => ({
      push: jest.fn(),
    }),
    useParams: () => ({
      address: testTokenAddress,
    }),
  };
});

describe('TokenAllowancePage', () => {
  const props = {
    origin: 'https://metamask.github.io',
    siteImage: 'https://metamask.github.io/test-dapp/metamask-fox.svg',
    useNonceField: false,
    currentCurrency: 'usd',
    nativeCurrency: 'GoerliETH',
    ethTransactionTotal: '0.0012',
    fiatTransactionTotal: '1.6',
    hexTransactionTotal: '0x44364c5bb0000',
    supportsEIP1559: true,
    userAddress: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
    tokenAddress: '0x55797717b9947b31306f4aac7ad1365c6e3923bd',
    data: '0x095ea7b30000000000000000000000009bc5baf874d2da8d216ae9f137804184ee5afef40000000000000000000000000000000000000000000000000000000000011170',
    isSetApproveForAll: false,
    setApproveForAllArg: false,
    decimals: '4',
    dappProposedTokenAmount: '7',
    currentTokenBalance: '10',
    toAddress: '0x9bc5baf874d2da8d216ae9f137804184ee5afef4',
    tokenSymbol: 'TST',
    showCustomizeGasModal: jest.fn(),
    warning: '',
    txData: {
      id: 3049568294499567,
      time: 1664449552289,
      status: 'unapproved',
      originalGasEstimate: '0xea60',
      userEditedGasLimit: false,
      chainId: CHAIN_IDS.MAINNET,
      loadingDefaults: false,
      dappSuggestedGasFees: {
        gasPrice: '0x4a817c800',
        gas: '0xea60',
      },
      sendFlowHistory: [],
      txParams: {
        from: '0x0dcd5d886577d5081b0c52e242ef29e70be3e7bc',
        to: '0x55797717b9947b31306f4aac7ad1365c6e3923bd',
        value: '0x0',
        data: '0x095ea7b30000000000000000000000009bc5baf874d2da8d216ae9f137804184ee5afef40000000000000000000000000000000000000000000000000000000000011170',
        gas: '0xea60',
        gasPrice: '0x4a817c800',
        maxFeePerGas: '0x4a817c800',
      },
      origin: 'https://metamask.github.io',
      type: 'approve',
      userFeeLevel: 'custom',
      defaultGasEstimates: {
        estimateType: 'custom',
        gas: '0xea60',
        maxFeePerGas: '0x4a817c800',
        maxPriorityFeePerGas: '0x4a817c800',
        gasPrice: '0x4a817c800',
      },
    },
  };

  let store;
  beforeEach(() => {
    store = configureMockStore([thunk])(state);
  });

  describe('when mounted', () => {
    it('should match snapshot', () => {
      const { container } = renderWithProvider(
        <TokenAllowance {...props} />,
        store,
      );

      expect(container).toMatchSnapshot();
    });

    it('should load the page with dappProposedAmount prefilled and "Use site suggestion" should not be displayed', () => {
      mockedState.appState.customTokenAmount = '';

      const { queryByText, getByTestId } = renderWithProvider(
        <TokenAllowance {...props} />,
        configureMockStore([thunk])(mockedState),
      );

      const useSiteSuggestion = queryByText('Use site suggestion');
      expect(useSiteSuggestion).not.toBeInTheDocument();

      const input = getByTestId('custom-spending-cap-input');
      expect(input.value).toBe('7');
    });
  });

  it('should render title "Spending cap request for your" in token allowance page', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    expect(getByText('Spending cap request for your')).toBeInTheDocument();
  });

  it('should render reject button', () => {
    const { getByTestId } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    const onCloseBtn = getByTestId('page-container-footer-cancel');
    expect(onCloseBtn).toBeInTheDocument();
  });

  it('should not render customize nonce modal if useNonceField is set to false', () => {
    const { queryByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    expect(queryByText('Nonce')).not.toBeInTheDocument();
    expect(queryByText('1')).not.toBeInTheDocument();
    expect(mockShowModal).not.toHaveBeenCalledTimes(1);
  });

  it('should render customize nonce modal if useNonceField is set to true', () => {
    props.useNonceField = true;
    props.nextNonce = 1;
    const { queryByText, getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    const editButton = getByText('Edit');
    expect(queryByText('Nonce')).toBeInTheDocument();
    expect(queryByText('1')).toBeInTheDocument();
    fireEvent.click(editButton);
    expect(mockShowModal).toHaveBeenCalledTimes(1);
  });

  it('should render nextNonce value when custom nonce value is a empty string', () => {
    props.useNonceField = true;
    props.customNonceValue = '';
    const { queryByText, getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    const editButton = getByText('Edit');
    expect(queryByText('Nonce')).toBeInTheDocument();
    expect(queryByText('1')).toBeInTheDocument();
    fireEvent.click(editButton);
    expect(mockShowModal).toHaveBeenCalledTimes(2);
  });

  it('should render edited custom nonce value', () => {
    props.useNonceField = true;
    state.metamask.customNonceValue = '3';
    const { queryByText, getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    const editButton = getByText('Edit');
    expect(queryByText('Nonce')).toBeInTheDocument();
    expect(queryByText('3')).toBeInTheDocument();
    fireEvent.click(editButton);
    expect(mockShowModal).toHaveBeenCalledTimes(3);
  });

  it('should render customize nonce warning if custom nonce value is higher than nextNonce value', () => {
    props.useNonceField = true;
    props.nextNonce = 2;
    props.customNonceValue = '3';
    props.warning = 'Nonce is higher than suggested nonce of 2';
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    expect(
      getByText('Nonce is higher than suggested nonce of 2'),
    ).toBeInTheDocument();
  });

  it('should not render customize nonce warning if custom nonce value is lower than nextNonce value', () => {
    props.useNonceField = true;
    props.nextNonce = 2;
    props.customNonceValue = '1';
    props.warning = '';
    const { container } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );
    const customizeNonceWarning = container.querySelector(
      '.token-allowance-container__custom-nonce-warning',
    );
    expect(customizeNonceWarning).not.toBeInTheDocument();
  });

  it('should render customize nonce modal when next button is clicked and if useNonceField is set to true', () => {
    props.useNonceField = true;
    state.metamask.customNonceValue = '2';
    const { getByText, getAllByText, queryByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    const nextButton = getByText('Next');
    fireEvent.click(nextButton);

    const editButton = getAllByText('Edit');
    expect(queryByText('Nonce')).toBeInTheDocument();
    expect(queryByText('2')).toBeInTheDocument();
    fireEvent.click(editButton[1]);
    expect(mockShowModal).toHaveBeenCalledTimes(4);
  });

  it('should render customize nonce modal when next button is clicked, than back button is clicked, than return to previous page and if useNonceField is set to true', () => {
    props.useNonceField = true;
    state.metamask.customNonceValue = '2';
    const { getByText, queryByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    const nextButton = getByText('Next');
    fireEvent.click(nextButton);

    const backButton = getByText('< Back');
    fireEvent.click(backButton);

    const editButton = getByText('Edit');
    expect(queryByText('Nonce')).toBeInTheDocument();
    expect(queryByText('2')).toBeInTheDocument();
    fireEvent.click(editButton);
    expect(mockShowModal).toHaveBeenCalledTimes(5);
  });

  it('should click View details and show function type', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    const viewDetailsButton = getByText('View details');
    fireEvent.click(viewDetailsButton);
    expect(getByText('Function: Approve')).toBeInTheDocument();
  });

  it('should click "Use site suggestion" and set input value to dappProposedTokenAmount', () => {
    mockedState.appState.customTokenAmount = '';

    const { getByText, getByTestId } = renderWithProvider(
      <TokenAllowance {...props} />,
      configureMockStore([thunk])(mockedState),
    );
    const textField = getByTestId('custom-spending-cap-input');

    expect(textField.value).toBe('7');
    fireEvent.change(textField, { target: { value: '1' } });
    expect(textField.value).toBe('1');

    const useSiteSuggestion = getByText('Use site suggestion');
    expect(useSiteSuggestion).toBeInTheDocument();

    act(() => {
      fireEvent.click(useSiteSuggestion);
    });

    expect(textField.value).toBe('7');
  });

  it('should call back button when button is clicked and return to previous page', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    const nextButton = getByText('Next');
    fireEvent.click(nextButton);

    expect(getByText('Site requested spending cap')).toBeInTheDocument();

    const backButton = getByText('< Back');
    fireEvent.click(backButton);

    expect(getByText('Spending cap request for your')).toBeInTheDocument();
  });

  it('should click Verify third-party details and show popup Third-party details, then close popup', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    const verifyThirdPartyDetails = getByText('Verify third-party details');
    fireEvent.click(verifyThirdPartyDetails);

    expect(getByText('Third-party details')).toBeInTheDocument();

    const gotIt = getByText('Got it');
    fireEvent.click(gotIt);
    expect(gotIt).not.toBeInTheDocument();
  });

  it('should show ledger info text if the sending address is ledger', () => {
    const { queryByText, getByText } = renderWithProvider(
      <TokenAllowance {...props} fromAddressIsLedger />,
      store,
    );

    expect(queryByText('Prior to clicking confirm:')).toBeNull();

    const nextButton = getByText('Next');
    fireEvent.click(nextButton);

    expect(queryByText('Prior to clicking confirm:')).toBeInTheDocument();
  });

  it('should not show ledger info text if the sending address is not ledger', () => {
    const { queryByText, getByText } = renderWithProvider(
      <TokenAllowance {...props} fromAddressIsLedger={false} />,
      store,
    );

    expect(queryByText('Prior to clicking confirm:')).toBeNull();

    const nextButton = getByText('Next');
    fireEvent.click(nextButton);

    expect(queryByText('Prior to clicking confirm:')).toBeNull();
  });

  it('should render security provider response if transaction is malicious', () => {
    const securityProviderResponse = {
      flagAsDangerous: 1,
      reason:
        'This has been flagged as potentially suspicious. If you sign, you could lose access to all of your NFTs and any funds or other assets in your wallet.',
      reason_header: 'Warning',
    };
    const { getByText } = renderWithProvider(
      <TokenAllowance
        {...props}
        txData={{ ...props.txData, securityProviderResponse }}
      />,
      store,
    );

    expect(getByText(securityProviderResponse.reason)).toBeInTheDocument();
  });

  it('should render from account name in header', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      store,
    );

    expect(getByText('Account 1')).toBeInTheDocument();
  });

  it('should account name from transaction even if currently selected account is different', () => {
    const newState = {
      ...state,
      metamask: {
        ...state.metamask,
      },
    };
    const newStore = configureMockStore([thunk])(newState);
    const { queryByText } = renderWithProvider(
      <TokenAllowance {...props} />,
      newStore,
    );

    expect(queryByText('Account 1')).toBeInTheDocument();
    expect(queryByText('Account 2')).not.toBeInTheDocument();
  });

  it('should display security alert if present', () => {
    const { getByText } = renderWithProvider(
      <TokenAllowance
        {...props}
        txData={{
          ...props.txData,
          securityAlertResponse: {
            resultType: 'Malicious',
            reason: 'blur_farming',
            description:
              'A SetApprovalForAll request was made on {contract}. We found the operator {operator} to be malicious',
            args: {
              contract: '0xa7206d878c5c3871826dfdb42191c49b1d11f466',
              operator: '0x92a3b9773b1763efa556f55ccbeb20441962d9b2',
            },
          },
        }}
      />,
      store,
    );

    expect(getByText('This is a deceptive request')).toBeInTheDocument();
  });
});
