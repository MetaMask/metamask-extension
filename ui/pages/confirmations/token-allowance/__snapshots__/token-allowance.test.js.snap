// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenAllowancePage when mounted should match snapshot 1`] = `
<div>
  <div
    class="box token-allowance-container page-container box--flex-direction-row"
  >
    <div
      class="box box--flex-direction-row"
    >
      <div
        class="confirm-page-container-navigation"
        style="display: none;"
      >
        <div
          class="confirm-page-container-navigation__container"
          data-testid="navigation-container"
          style="visibility: hidden;"
        >
          <button
            class="confirm-page-container-navigation__arrow"
            data-testid="first-page"
          >
            <i
              class="fa fa-angle-double-left fa-2x"
            />
          </button>
          <button
            class="confirm-page-container-navigation__arrow"
            data-testid="previous-page"
          >
            <i
              class="fa fa-angle-left fa-2x"
            />
          </button>
        </div>
        <div
          class="confirm-page-container-navigation__textcontainer"
        >
          <div
            class="confirm-page-container-navigation__navtext"
          >
            0
             
            of
             
            0
          </div>
          <div
            class="confirm-page-container-navigation__longtext"
          >
            requests waiting to be acknowledged
          </div>
        </div>
        <div
          class="confirm-page-container-navigation__container"
          style="visibility: hidden;"
        >
          <button
            class="confirm-page-container-navigation__arrow"
            data-testid="next-page"
          >
            <i
              class="fa fa-angle-right fa-2x"
            />
          </button>
          <button
            class="confirm-page-container-navigation__arrow"
            data-testid="last-page"
          >
            <i
              class="fa fa-angle-double-right fa-2x"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="box box--padding-right-4 box--padding-left-4 box--display-flex box--flex-direction-row box--justify-content-space-between box--align-items-center"
    >
      <div
        class="box box--flex-direction-row"
      />
      <div
        class="box box--flex-direction-row box--text-align-end"
      >
        <h6
          class="mm-box mm-text mm-text--body-sm mm-text--font-weight-bold mm-box--color-text-muted"
        >
          1
           
          of
           2
        </h6>
      </div>
    </div>
    <div
      class="box network-account-balance-header box--padding-4 box--display-flex box--flex-direction-row box--justify-content-space-between box--align-items-center"
    >
      <div
        class="box box--display-flex box--gap-2 box--flex-direction-row box--align-items-center"
      >
        <div
          class="box box--display-flex box--flex-direction-row box--align-items-center"
        >
          <div
            class=""
          >
            <div
              class="identicon"
              style="height: 32px; width: 32px; border-radius: 16px;"
            >
              <div
                style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 32px; height: 32px; display: inline-block; background: rgb(250, 58, 0);"
              >
                <svg
                  height="32"
                  width="32"
                  x="0"
                  y="0"
                >
                  <rect
                    fill="#18CDF2"
                    height="32"
                    transform="translate(-1.04839350379394 -3.3042840694604987) rotate(328.9 16 16)"
                    width="32"
                    x="0"
                    y="0"
                  />
                  <rect
                    fill="#035E56"
                    height="32"
                    transform="translate(-18.298461708832043 10.5924618717486) rotate(176.2 16 16)"
                    width="32"
                    x="0"
                    y="0"
                  />
                  <rect
                    fill="#F26602"
                    height="32"
                    transform="translate(16.667842018223922 -14.205139722997082) rotate(468.9 16 16)"
                    width="32"
                    x="0"
                    y="0"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="network-account-balance-header__network-account__ident-icon-ethereum"
            style="height: 16px; width: 16px;"
          >
            <img
              alt="Ethereum Mainnet"
              src="./images/eth_logo.svg"
              style="height: 16px; width: 16px;"
            />
          </div>
        </div>
        <div
          class="box box--display-flex box--flex-direction-column box--align-items-flex-start"
        >
          <h6
            class="mm-box mm-text mm-text--body-sm mm-box--color-text-alternative"
            data-testid="signature-request-network-display"
          >
            Ethereum Mainnet
          </h6>
          <h6
            class="mm-box mm-text mm-text--body-sm mm-text--font-weight-bold mm-box--color-text-default"
          >
            Account 1
          </h6>
        </div>
      </div>
      <div
        class="box box--display-flex box--flex-direction-column box--align-items-flex-end"
      >
        <h6
          class="mm-box mm-text mm-text--body-sm mm-box--color-text-alternative"
        >
          Balance
        </h6>
        <h6
          align="end"
          class="mm-box mm-text mm-text--body-sm mm-text--font-weight-bold mm-box--color-text-default"
        >
          10
           
          TST
        </h6>
      </div>
    </div>
    
    <div
      class="box box--display-flex box--flex-direction-row box--justify-content-center"
    >
      <div
        class="box token-allowance-container__icon-display-content box--margin-top-6 box--margin-right-12 box--margin-bottom-8 box--margin-left-12 box--padding-top-2 box--padding-right-4 box--padding-bottom-2 box--padding-left-2 box--display-flex box--flex-direction-row box--align-items-center box--border-style-solid box--border-color-border-muted box--border-width-1"
      >
        <div
          class=""
        >
          <img
            alt="https://metamask.github.io"
            class="url-icon token-allowance-container__icon-display-content__siteimage-identicon"
            src="https://metamask.github.io/test-dapp/metamask-fox.svg"
          />
        </div>
        <h6
          class="mm-box mm-text mm-text--body-sm mm-box--margin-left-1 mm-box--color-text-alternative"
        >
          https://metamask.github.io
        </h6>
      </div>
    </div>
    <div
      class="box box--margin-right-4 box--margin-left-4 box--flex-direction-row"
    >
      <h3
        align="center"
        class="mm-box mm-text mm-text--heading-md mm-box--color-text-default"
      >
        <span>
           
          Spending cap request for your 
          <div
            class="box box--margin-top-4 box--flex-direction-row"
          >
            <div
              class="box contract-token-values box--display-flex box--gap-2 box--flex-direction-row box--justify-content-center box--align-items-center"
            >
              <div
                class=""
              >
                <div
                  class="identicon"
                  style="height: 24px; width: 24px; border-radius: 12px;"
                >
                  <div
                    style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 24px; height: 24px; display: inline-block; background: rgb(242, 206, 2);"
                  >
                    <svg
                      height="24"
                      width="24"
                      x="0"
                      y="0"
                    >
                      <rect
                        fill="#186FF2"
                        height="24"
                        transform="translate(4.5535541527659795 -0.012406777909352561) rotate(501.9 12 12)"
                        width="24"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#F97101"
                        height="24"
                        transform="translate(-10.003810991077078 6.326210026231173) rotate(323.7 12 12)"
                        width="24"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#FB1832"
                        height="24"
                        transform="translate(-22.151479903389234 -4.322407331572157) rotate(370.5 12 12)"
                        width="24"
                        x="0"
                        y="0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <h2
                class="mm-box mm-text mm-text--heading-lg mm-text--font-weight-bold mm-box--margin-top-0 mm-box--margin-bottom-0 mm-box--color-text-alternative"
              >
                TST
              </h2>
              <div>
                <div
                  aria-describedby="tippy-tooltip-1"
                  class=""
                  data-original-title="Copy to clipboard"
                  data-tooltipped=""
                  style="display: inline;"
                  tabindex="0"
                >
                  <button
                    aria-label="Copy to clipboard"
                    class="mm-box mm-button-icon mm-button-icon--size-lg mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-muted mm-box--background-color-transparent mm-box--rounded-lg"
                  >
                    <span
                      class="mm-box mm-icon mm-icon--size-lg mm-box--display-inline-block mm-box--color-inherit"
                      style="mask-image: url('./images/icons/copy.svg');"
                    />
                  </button>
                </div>
              </div>
              <div>
                <div
                  aria-describedby="tippy-tooltip-2"
                  class=""
                  data-original-title="Open in block explorer"
                  data-tooltipped=""
                  style="display: inline;"
                  tabindex="0"
                >
                  <button
                    aria-label="Open in block explorer"
                    class="mm-box mm-button-icon mm-button-icon--size-lg mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-muted mm-box--background-color-transparent mm-box--rounded-lg"
                  >
                    <span
                      class="mm-box mm-icon mm-icon--size-lg mm-box--display-inline-block mm-box--color-inherit"
                      style="mask-image: url('./images/icons/export.svg');"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          
           
        </span>
      </h3>
    </div>
    <div
      class="box box--margin-top-1 box--display-flex box--flex-direction-row box--justify-content-center"
    >
      <a
        class="button btn-link token-allowance-container__verify-link"
        role="button"
        tabindex="0"
      >
        <h6
          class="mm-box mm-text mm-text--body-sm mm-box--color-primary-default"
        >
          Verify third-party details
        </h6>
      </a>
    </div>
    <div
      class="box box--margin-4 box--sm:margin-4 box--md:margin-3 box--lg:margin-4 box--flex-direction-row"
    >
      <div
        class="mm-box custom-spending-cap mm-box--padding-4 mm-box--display-flex mm-box--gap-2 mm-box--flex-direction-column mm-box--align-items-flex-start mm-box--background-color-background-alternative mm-box--rounded-sm"
      >
        <div
          class="mm-box mm-box--display-flex mm-box--gap-1 mm-box--align-items-center mm-box--width-full"
        >
          <div>
            <label
              class="mm-box mm-text mm-label mm-label--html-for mm-text--body-md mm-text--font-weight-medium mm-box--margin-right-1 mm-box--display-inline mm-box--align-items-center mm-box--color-text-default"
              for="custom-spending-cap"
            >
              Custom spending cap
            </label>
            <div
              class="mm-box mm-box--display-inline-flex mm-box--align-items-center"
            >
              <div>
                <div
                  aria-describedby="tippy-tooltip-3"
                  class=""
                  data-original-title="null"
                  data-tooltipped=""
                  style="display: inline;"
                  tabindex="0"
                >
                  <span
                    class="mm-box mm-icon mm-icon--size-inherit mm-box--display-inline-block mm-box--color-icon-alternative"
                    style="mask-image: url('./images/icons/question.svg');"
                  />
                </div>
              </div>
            </div>
          </div>
          <button
            class="mm-box mm-text mm-button-base mm-button-link mm-button-link--size-auto mm-text--body-md-medium mm-box--margin-left-auto mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
          >
            Use site suggestion
          </button>
        </div>
        <div
          class="mm-box mm-text-field mm-text-field--size-md mm-text-field--focused mm-text-field--truncate mm-box--padding-right-4 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--width-full mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
        >
          <input
            autocomplete="off"
            class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-md mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-2 mm-box--padding-left-4 mm-box--padding-inline-end-4 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
            data-testid="custom-spending-cap-input"
            focused="true"
            id="custom-spending-cap"
            placeholder="Enter a number"
            type="text"
            value="1"
          />
          <button
            class="mm-box mm-text mm-button-base mm-button-link mm-button-link--size-auto mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
            data-testid="custom-spending-cap-max-button"
          >
            Max
          </button>
        </div>
        <div
          class="mm-box mm-text mm-help-text mm-text--body-sm mm-box--color-text-default"
        >
          <span>
             
            This allows the third party to spend 
            <span
              class="mm-box mm-text mm-text--body-sm-bold mm-box--color-text-default"
            >
              1
               
              TST
            </span>
             from your current balance.
             
          </span>
        </div>
        <a
          class="mm-box mm-text mm-button-base mm-button-base--size-sm mm-button-link mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
          href="https://support.metamask.io/privacy-and-security/how-to-customize-token-approvals-with-a-spending-cap/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Learn more
        </a>
      </div>
    </div>
    <div
      class="box box--margin-right-4 box--margin-left-4 box--flex-direction-row"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-center mm-box--align-items-center"
      />
    </div>
    <div
      class="box box--display-flex box--flex-direction-row box--justify-content-center"
    >
      <a
        class="button btn-link token-allowance-container__view-details"
        role="button"
        tabindex="0"
      >
        <h6
          class="mm-box mm-text mm-text--body-sm mm-box--margin-right-1 mm-box--color-primary-default"
        >
          View details
        </h6>
        <i
          class="fa fa-sm fa-angle-down"
        />
      </a>
    </div>
    <div
      class="page-container__footer"
    >
      <footer>
        <button
          class="button btn--rounded btn-secondary page-container__footer-button page-container__footer-button__cancel"
          data-testid="page-container-footer-cancel"
        >
          Reject
        </button>
        <button
          class="button btn--rounded btn-primary page-container__footer-button"
          data-testid="page-container-footer-next"
        >
          Next
        </button>
      </footer>
    </div>
  </div>
</div>
`;
